cmd/ipfs/kubo/start.go

```diff
@@ -0,0 +1,484 @@
+)
+
+// log is the command logger.
+var (
+	log    = logging.Logger("cmd/ipfs")
+	tracer trace.Tracer
+)
+
+// declared as a var for testing purposes.
+var dnsResolver = madns.DefaultResolver
+
+const (
+	EnvEnableProfiling = "IPFS_PROF"
+	cpuProfile         = "ipfs.cpuprof"
+	heapProfile        = "ipfs.memprof"
+)
+
+type PluginPreloader func(*loader.PluginLoader) error
+
+func LoadPlugins(repoPath string, preload PluginPreloader) (*loader.PluginLoader, error) {
+	plugins, err := loader.NewPluginLoader(repoPath)
+	if err != nil {
+		return nil, fmt.Errorf("error loading plugins: %s", err)
+	}
+
+	if preload != nil {
+		if err := preload(plugins); err != nil {
+			return nil, fmt.Errorf("error loading plugins (preload): %s", err)
+		}
+	}
+
+	if err := plugins.Initialize(); err != nil {
+		return nil, fmt.Errorf("error initializing plugins: %s", err)
+	}
+
+	if err := plugins.Inject(); err != nil {
+		return nil, fmt.Errorf("error initializing plugins: %s", err)
+	}
+	return plugins, nil
+}
+
+func printErr(err error) int {
+	fmt.Fprintf(os.Stderr, "Error: %s\n", err.Error())
+	return 1
+}
+
+func newUUID(key string) logging.Metadata {
+	ids := "#UUID-ERROR#"
+	if id, err := uuid.NewRandom(); err == nil {
+		ids = id.String()
+	}
+	return logging.Metadata{
+		key: ids,
+	}
+}
+
+func BuildDefaultEnv(ctx context.Context, req *cmds.Request) (cmds.Environment, error) {
+	return BuildEnv(nil)(ctx, req)
+}
+
+func BuildEnv(pl PluginPreloader) func(ctx context.Context, req *cmds.Request) (cmds.Environment, error) {
+	return func(ctx context.Context, req *cmds.Request) (cmds.Environment, error) {
+		checkDebug(req)
+		repoPath, err := GetRepoPath(req)
+		if err != nil {
+			return nil, err
+		}
+		log.Debugf("config path is %s", repoPath)
+
+		plugins, err := LoadPlugins(repoPath, pl)
+		if err != nil {
+			return nil, err
+		}
+
+		// this sets up the function that will initialize the node
+		// this is so that we can construct the node lazily.
+		return &oldcmds.Context{
+			ConfigRoot: repoPath,
+			ReqLog:     &oldcmds.ReqLog{},
+			Plugins:    plugins,
+			ConstructNode: func() (n *core.IpfsNode, err error) {
+				if req == nil {
+					return nil, errors.New("constructing node without a request")
+				}
+
+				r, err := fsrepo.Open(repoPath)
+				if err != nil { // repo is owned by the node
+					return nil, err
+				}
+
+				// ok everything is good. set it on the invocation (for ownership)
+				// and return it.
+				n, err = core.NewNode(ctx, &core.BuildCfg{
+					Repo: r,
+				})
+				if err != nil {
+					return nil, err
+				}
+
+				return n, nil
+			},
+		}, nil
+	}
+}
+
+// Start roadmap:
+// - parse the commandline to get a cmdInvocation
+// - if user requests help, print it and exit.
+// - run the command invocation
+// - output the response
+// - if anything fails, print error, maybe with help.
+func Start(buildEnv func(ctx context.Context, req *cmds.Request) (cmds.Environment, error)) (exitCode int) {
```

Comment:
To be honest, this feels like the wrong thing to export. We should just be exporting a way to start the daemon that's entirely independent of the commandline tool.

Reply:
Maybe, although I think for what @magik6k was looking for he wanted both the CLI and daemon and so this seems natural.

IIUC it's basically a way to do preloaded plugins with a fairly small and Go versioned codebase that uses imports rather than what effectively operates like a kubo fork.

For example, instead of:
- git pull the latest version, add your plugins as a dependency, compile
- update kubo dependency in your repo, compile

It's got some tradeoffs with regards to a standard preloaded plugin build, but they seem like reasonable ones for someone to make.

Were you thinking to expose two different things, one for starting the daemon and another for the CLI?

Reply:
In my case I want to have the whole CLI with the plugin, doing it with start meant the least amount of code needed to get the cli + plugin